docs/references/architecture/adr-115-predictable-block-times.md

```diff
@@ -0,0 +1,129 @@
+# ADR 115: Predictable Block Times
+
+## Changelog
+
+ - April 30 2024: Created by @melekes
+ - May 13 2024: Updated by @melekes
+
+## Status
+
+**Proposed**
+
+## Context
+
+Let's say you're an ABCI application developer and you want to have
+constant block times: 1 block each 6s. How would you do that?
+
+You found out that proposing a block and voting on it takes roughly 2s in your
+network. You instruct validators in your network to change `timeout_commit`
+(_how long a node waits for additional precommits after committing a block,
+before starting on the next height_) from 1s to 4s in the node's config file.
+
+Do you have predictable block times now?
+
+No - you don't. The expected block time will be around 6s, but even in
+favorable runs it will drift far apart from 6s due to:
+
+1. CometBFT going into multiple rounds of consensus (it happens rarely but
+   still).
+2. Network latency.
+3. Clock drifts.
+4. The delay for processing a block.
+
+A validator could also change `timeout_commit` to 0s in its node's config file.
+That means whenever THIS validator proposes a block, the block time will be 2s
+(not 6s!).
+
+Note that the value of `timeout_commit` is static and can't be changed
+without restarting the node.
+
+Because 1-3 are out of your (and our) control, **we can't have constant block
+times**. But we can design a mechanism so that the medium-to-long term average
+block time converges to a desired value.
+
+To achieve that, we need to define a form of variable block time. Namely, if a
+block takes longer than expected, we should be able to render the next block(s)
+faster in order to converge to the desired (average) block time.
+
+### Use Cases
+
+In case of Osmosis, committing a big block is expected to take longer than
+usual. If we know the approximate size of the next block, we could increase
+`timeout_commit` dynamically (if such feature existed) to give the state
+machine some extra time to finish execution.
+
+In case of Berachain, they want to have a constant block time to match
+Ethereum's slots, which are equal to 12s.
+
+## Proposal
+
+Move `timeout_commit` into `FinalizeBlockResponse` as `next_block_delay`. This
+field's semantics stays essentially the same: delay between the time when the
+current block is committed and the next height is started. The idea is
+literally to have the same behavior as `timeout_commit`, while allowing the
+application to pick a different value for each height.
```

Comment:
fwiw, if I'm interpreting this correctly we've tried this [here](https://github.com/celestiaorg/celestia-core/pull/965)

in our network tests, everything worked as expected and we have very consistent block times. However when we shipped this to a testnet, we frequently failed to reach consensus.

replicating and debugging this issue by changing the clocks of the network instances is next on our list of todos after we switch network testing backends https://github.com/celestiaorg/celestia-core/issues/1333

Reply:
Thanks for flagging this. We will keep this in mind when implementing this feature.